nibabel/pointset.py

```diff
@@ -144,6 +149,77 @@ def get_coords(self, *, as_homogeneous: bool = False):
         return coords
 
 
+@dataclass(init=False)
+class TriangularMesh(Pointset):
+    triangles: CoordinateArray
+
+    def __init__(
+        self,
+        coordinates: CoordinateArray,
+        triangles: CoordinateArray,
+        affine: np.ndarray | None = None,
+        homogeneous: bool = False,
+    ):
+        super().__init__(coordinates, affine=affine, homogeneous=homogeneous)
+        self.triangles = triangles
+
+    @classmethod
+    def from_tuple(
+        cls,
+        mesh: tuple[CoordinateArray, CoordinateArray],
+        affine: np.ndarray | None = None,
+        homogeneous: bool = False,
+        **kwargs,
+    ) -> Self:
+        return cls(mesh[0], mesh[1], affine=affine, homogeneous=homogeneous, **kwargs)
+
+    @classmethod
+    def from_object(
+        cls,
+        mesh: HasMeshAttrs,
+        affine: np.ndarray | None = None,
+        homogeneous: bool = False,
+        **kwargs,
+    ) -> Self:
+        return cls(
+            mesh.coordinates, mesh.triangles, affine=affine, homogeneous=homogeneous, **kwargs
+        )
+
+    @property
+    def n_triangles(self):
+        """Number of faces
+
+        Subclasses should override with more efficient implementations.
+        """
+        return self.triangles.shape[0]
+
+    def get_triangles(self):
+        """Mx3 array of indices into coordinate table"""
+        return np.asanyarray(self.triangles)
+
+    def get_mesh(self, *, as_homogeneous: bool = False):
+        return self.get_coords(as_homogeneous=as_homogeneous), self.get_triangles()
+
+
+class CoordinateFamilyMixin(Pointset):
+    def __init__(self, *args, name='original', **kwargs):
+        mapping = kwargs.pop('mapping', {})
+        super().__init__(*args, **kwargs)
+        self._coords = {name: self.coordinates, **mapping}
+
+    def get_names(self):
+        """List of surface names that can be passed to :meth:`with_name`"""
+        return list(self._coords)
+
+    def with_name(self, name: str) -> Self:
+        new = replace(self, coordinates=self._coords[name])
+        new._coords = self._coords
+        return new
+
+    def add_coordinates(self, name, coordinates):
+        self._coords[name] = coordinates
```

Comment:
This is simple but perhaps a little clunky. We could also do:

```suggestion
    def add_coordinates(self, *args, **kwargs):
        self._coords.update(dict(*args, **kwargs))
```

This would allow it to take any inputs that would generate a `dict`, which might be a bit more ergonomic. I'm not 100% on this, so I haven't made the change.